feature-antikler-BJS2-65238

antikler opened the pull request.

antikler committed:
BSJ2-65238. What made:
1) added LikeController
2) added LikeDto
3) added LikeMapper
4) added LikeService and first implementations LikeServiceImpl

antikler committed:
BSJ2-65238. What made:
1) added LikeController
2) added LikeDto
3) added LikeMapper
4) added LikeService and first implementations LikeServiceImpl

Commit:
BJS2-65238 Система лайков:
task done without tests

Commit:
BJS2-65238 Система лайков:
task done without tests

Commit:
Merge remote-tracking branch 'origin/feature-antikler-BJS2-65238' into feature-antikler-BJS2-65238

Commit:
BJS2-65238 Система лайков:
task is done

Valerik1984 commented on src/main/java/faang/school/postservice/dto/like/LikeDto.java:
```diff
@@ -0,0 +1,21 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class LikeDto {
+    @NotNull
```
Зачем @NotNull на примитиве?

antikler replied:
это я психанул. Исправил.

Valerik1984 commented on src/main/java/faang/school/postservice/controller/like/LikeController.java:
```diff
@@ -0,0 +1,46 @@
+    private final LikeService likeService;
+
+    @PostMapping("/posts/{postId}/like")
+    public ResponseEntity<LikeDto> likePost(@PathVariable("postId") long postId,
```
Нет первоначальной валидации на 0 и на отрицательные значения! Так везде.
Пример. if (postId <= 0 || headerUserId <= 0) {
return ResponseEntity.badRequest().body("Invalid postId or userId");
}

antikler replied:
Исправил через аннотации

Valerik1984 commented on src/main/java/faang/school/postservice/service/like/implementations/LikeServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    private final CommentRepository commentRepository;
+
+    @Override
+    public LikeDto likePost(long postId, long userId) {
```
Отсутствуют аннотации @Transactional.

antikler replied:
Согласен. Исправил.

Valerik1984 commented on src/main/java/faang/school/postservice/service/like/implementations/LikeServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    public LikeDto likePost(long postId, long userId) {
+        Post post = checkPostId(postId);
+        checkAuthor(userId);
+        likeRepository.findByPostIdAndUserId(postId, userId)
```
Нет проверки на одновременность лайков в посте и комментарии

likeRepository.findByUserId(userId)
.ifPresent(like -> {
throw new LikeAlreadyExistException(
String.format("User %d already has a like (postId=%d or commentId=%d)",
userId, like.getPost() != null ? like.getPost().getId() : 0,
like.getComment() != null ? like.getComment().getId() : 0));
});

antikler replied:
Не согласен.
Эндпоинты разнесены для отдельно для лайков комментариев и отдельно для постов. 
commentId и postId не могут фигурировать одновременно в одном запросе.
В условии задачи стоит ограничение установки лайка именно одновременно и для поста и для комментария.

Valerik1984 commented on src/main/java/faang/school/postservice/service/like/implementations/LikeServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    private void checkAuthor(long userId) {
+        try {
+            userServiceClient.getUser(userId);
+        } catch (FeignException e) {
```
Добавить проверку статуса } catch (FeignException e) { if (e.status() == 404) { throw new AuthorNotFoundException("Author with id " + userId + " not found"); } log.error("Failed to check author: id={}", userId, e); throw new RuntimeException("User service unavailable", e); }

antikler replied:
согласен. Исправил.

Valerik1984 commented on src/main/java/faang/school/postservice/service/like/implementations/LikeServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Override
+    public LikeDto likeComment(long commentId, long userId) {
```
Содержат похожую логику с likePost, можно подумать, как оптимизировать.

Valerik1984 commented on src/main/java/faang/school/postservice/service/like/implementations/LikeServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+
+    @Override
+    public LikeDto likePost(long postId, long userId) {
+        Post post = checkPostId(postId);
```
Нет проверки статуса isDeleted у поста, если найден
Post post = checkPostId(postId);
if (post.isDeleted()) {
throw new PostNotFoundException("Post with ID " + postId + " is deleted");
}

antikler replied:
согласен, исправил.

Valerik1984 commented on src/main/java/faang/school/postservice/mapper/like/LikeMapper.java:
```diff
@@ -0,0 +1,18 @@
+import org.mapstruct.ReportingPolicy;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface LikeMapper {
```
не указано, как мапить commentId и postId из LikeDto в объекты Post и Comment в Like. MapStruct не сможет автоматически создать объекты Post и Comment из примитивов long.
Нужно явно указать маппинг или использовать зависимости для загрузки сущностей
Можно использовать декоратор, или внедрить репозитории через @Context

antikler replied:
сначала решал задачу через передачу LikeDto, но потом всё переделал.
Не стал пока реализовывать, да и не используется она у меня. Добавил маппинг.

Commit:
BJS2-65238 Система лайков:
fix bugs and refactoring code

antikler committed:
Merge branch 'manticore-master-stream9' into feature-antikler-BJS2-65238

Valerik1984 reviewed (approved):
Initial check passed, mentor check required!

antikler committed:
BJS2-65238 Система лайков:
fix bugs after code-review.
+ added @RequestHeader as required param in LikeController

Soddik commented on src/main/java/faang/school/postservice/dto/like/LikeDto.java:
```diff
@@ -0,0 +1,19 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class LikeDto {
+    long userId;
```
private?

antikler replied:
исправил.

Soddik commented on src/main/java/faang/school/postservice/dto/like/LikeDto.java:
```diff
@@ -0,0 +1,19 @@
+@NoArgsConstructor
+public class LikeDto {
+    long userId;
+    long commentId;
```
private?

antikler replied:
исправил.

Soddik commented on src/main/java/faang/school/postservice/dto/like/LikeDto.java:
```diff
@@ -0,0 +1,19 @@
+public class LikeDto {
+    long userId;
+    long commentId;
+    long postId;
```
private?

antikler replied:
исправил.

Soddik commented on src/main/java/faang/school/postservice/dto/like/LikeDto.java:
```diff
@@ -0,0 +1,19 @@
+    long userId;
+    long commentId;
+    long postId;
+    LocalDateTime createdAt;
```
private?

antikler replied:
исправил.

Soddik commented on src/main/java/faang/school/postservice/dto/comment/CommentDto.java:
```diff
@@ -18,4 +18,5 @@ public class CommentDto {
     @NotNull
     Long postId;
     LocalDateTime createdAt;
+    Long likes;
```
модификатор доступа private?

antikler replied:
исправил.

Soddik reviewed (changes requested):
Конфликт

Commit:
Merge branch 'manticore-master-stream9' into feature-antikler-BJS2-65238

# Conflicts:
#	src/main/java/faang/school/postservice/exception/GlobalExceptionHandler.java
#	src/main/java/faang/school/postservice/service/post/implementations/PostServiceImpl.java

Commit:
BJS2-65238 Система лайков:
conflict resolve
